### ISSUE TITLE
[QUESTION] The implementation of tensor_parallel.cross_entropy when using tensor parallel

### ISSUE THREAD

**Issue by robin087** [NVIDIA/Megatron-LM#445]
**Your question**
When using tensor parallel, we find that the cross_entropy loss computed by tensor_parallel.cross_entropy is different from torch.nn.functional.cross_entropy. The difference is about 1% for our data.
For the implementation of tensor_parallel.cross_entropy, we find that the loss is computed on the partition_vocab_size which is 8 times smaller than vocab_size (tp_size = 8). We think maybe this implementation causes the difference above.
In this case, is this implementation correct? Or can this implementation ensure the performance when using tensor parallel?

**Comment by rmwor1d** [comment]
Hi, we are recently facing the similar problem of numerical differences between megatron and torch,  and thus I'm wondering have you compared your input Tensor of both cross_entropy in your case, is there any difference between the inputs? or, the layers or operations much before? The way we compare two Tensors is `(TensorA - TensorB).abs().sum()`

**Comment by wdykas** [comment]
@robin087 when you see these numerical differences between `torch.nn.functional.cross_entropy` and ` tensor_parallel.cross_entropy` are you using ColumnParallelLinear as inputs for both or do you use `torch.nn.Linear` with `torch.nn.functional.cross_entropy`?

**Comment by odelalleau** [comment]
I'm not sure how this is done in Megatron-LM, so this may be irrelevant, but if TP changes the vocabulary padding then this may cause some differences.
